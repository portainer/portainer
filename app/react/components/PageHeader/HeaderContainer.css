.row.header .meta .page {
  padding-top: 7px;
}

body.hamburg .row.header .meta {
  margin-left: 70px;
}

.row.header {
  min-height: 60px;
  background: var(--bg-row-header-color);
  margin-bottom: 15px;
}
.row.header > div:last-child {
  padding-right: 0;
}
.row.header .meta .page {
  font-size: 17px;
  padding-top: 11px;
}

.row.header .meta div {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
.row.header .login a {
  padding: 18px;
  display: block;
}
.row.header .user {
  min-width: 130px;
}
.row.header .user > .item {
  width: 65px;
  height: 60px;
  float: right;
  display: inline-block;
  text-align: center;
  vertical-align: middle;
}
.row.header .user > .item a {
  color: #919191;
  display: block;
}
.row.header .user > .item i {
  font-size: 20px;
  line-height: 55px;
}
.row.header .user > .item img {
  width: 40px;
  height: 40px;
  margin-top: 10px;
  border-radius: 2px;
}
.row.header .user > .item ul.dropdown-menu {
  border-radius: 2px;
  -webkit-box-shadow: 0 6px 12px rgba(0, 0, 0, 0.05);
  box-shadow: 0 6px 12px rgba(0, 0, 0, 0.05);
}
.row.header .user > .item ul.dropdown-menu .dropdown-header {
  text-align: center;
}
.row.header .user > .item ul.dropdown-menu li.link {
  text-align: left;
}
.row.header .user > .item ul.dropdown-menu li.link a {
  padding-left: 7px;
  padding-right: 7px;
}
.row.header .user > .item ul.dropdown-menu:before {
  position: absolute;
  top: -7px;
  right: 23px;
  display: inline-block;
  border-right: 7px solid transparent;
  border-bottom: 7px solid rgba(0, 0, 0, 0.2);
  border-left: 7px solid transparent;
  content: '';
}
.row.header .user > .item ul.dropdown-menu:after {
  position: absolute;
  top: -6px;
  right: 24px;
  display: inline-block;
  border-right: 6px solid transparent;
  border-bottom: 6px solid #ffffff;
  border-left: 6px solid transparent;
  content: '';
}

/*angular-loading-bar override*/
#loadingbar-placeholder {
  margin-bottom: 0;
  height: 3px;
}

#loading-bar .bar {
  position: relative;
  height: 3px;
  background: var(--blue-3);
}
/*!angular-loading-bar override*/
