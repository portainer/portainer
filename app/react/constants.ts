export const STACK_NAME_VALIDATION_REGEX = '^[-_a-z0-9]+$';

export const BROWSER_OS_PLATFORM = getOs();

function getOs() {
  const { userAgent } = navigator;

  if (userAgent.includes('Windows')) {
    return 'win';
  }

  return userAgent.includes('Mac') ? 'mac' : 'lin';
}
