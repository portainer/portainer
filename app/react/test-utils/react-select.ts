/** Simulate user events on react-select dropdowns
 *
 * taken from https://github.com/lokalise/react-select-event/blob/migrate-to-user-event/src/index.ts
 * until package is updated
 */

import {
  Matcher,
  findAllByText,
  findByText,
  waitFor,
} from '@testing-library/dom';
import userEvent from '@testing-library/user-event';

// find the react-select container from its input field ðŸ¤·
function getReactSelectContainerFromInput(input: HTMLElement): HTMLElement {
  return input.parentNode!.parentNode!.parentNode!.parentNode!
    .parentNode as HTMLElement;
}

type User = ReturnType<typeof userEvent.setup> | typeof userEvent;

type UserEventOptions = {
  user?: User;
};

/**
 * Utility for opening the select's dropdown menu.
 * @param {HTMLElement} input The input field (eg. `getByLabelText('The label')`)
 */
export async function openMenu(
  input: HTMLElement,
  { user = userEvent }: UserEventOptions = {}
) {
  await user.click(input);
  await user.type(input, '{ArrowDown}');
}

// type text in the input field
async function type(
  input: HTMLElement,
  text: string,
  { user }: Required<UserEventOptions>
) {
  await user.type(input, text);
}

// press the "clear" button, and reset various states
async function clear(
  clearButton: Element,
  { user }: Required<UserEventOptions>
) {
  await user.click(clearButton);
}

interface Config extends UserEventOptions {
  /** A container where the react-select dropdown gets rendered to.
   *  Useful when rendering the dropdown in a portal using `menuPortalTarget`.
   */
  container?: HTMLElement | (() => HTMLElement);
}

/**
 * Utility for selecting a value in a `react-select` dropdown.
 * @param {HTMLElement} input The input field (eg. `getByLabelText('The label')`)
 * @param {Matcher|Matcher[]} optionOrOptions The display name(s) for the option(s) to select
 * @param {Object} config Optional config options
 * @param {HTMLElement | (() => HTMLElement)} config.container A container for the react-select and its dropdown (defaults to the react-select container)
 *            Useful when rending the dropdown to a portal using react-select's `menuPortalTarget`.
 *            Can be specified as a function if it needs to be lazily evaluated.
 */
export async function select(
  input: HTMLElement,
  optionOrOptions: Matcher | Array<Matcher>,
  { user = userEvent, ...config }: Config = {}
) {
  const options = Array.isArray(optionOrOptions)
    ? optionOrOptions
    : [optionOrOptions];

  // Select the items we care about
  // eslint-disable-next-line no-restricted-syntax
  for (const option of options) {
    await openMenu(input, { user });

    let container;
    if (typeof config.container === 'function') {
      // when specified as a function, the container needs to be lazily evaluated, so
      // we have to wait for it to be visible:
      await waitFor(config.container);
      container = config.container();
    } else if (config.container) {
      container = config.container;
    } else {
      container = getReactSelectContainerFromInput(input);
    }

    // only consider visible, interactive elements
    const matchingElements = await findAllByText(container, option, {
      ignore: "[aria-live] *,[style*='visibility: hidden']",
    });

    // When the target option is already selected, the react-select display text
    // will also match the selector. In this case, the actual dropdown element is
    // positioned last in the DOM tree.
    const optionElement = matchingElements[matchingElements.length - 1];
    await user.click(optionElement);
  }
}

interface CreateConfig extends Config, UserEventOptions {
  createOptionText?: string | RegExp;
  waitForElement?: boolean;
}
/**
 * Utility for creating and selecting a value in a Creatable `react-select` dropdown.
 * @async
 * @param {HTMLElement} input The input field (eg. `getByLabelText('The label')`)
 * @param {String} option The display name for the option to type and select
 * @param {Object} config Optional config options
 * @param {HTMLElement} config.container A container for the react-select and its dropdown (defaults to the react-select container)
 *                         Useful when rending the dropdown to a portal using react-select's `menuPortalTarget`
 * @param {boolean} config.waitForElement Whether create should wait for new option to be populated in the select container
 * @param {String|RegExp} config.createOptionText Custom label for the "create new ..." option in the menu (string or regexp)
 */
export async function create(
  input: HTMLElement,
  option: string,
  { waitForElement = true, user = userEvent, ...config }: CreateConfig = {}
) {
  const createOptionText = config.createOptionText || /^Create "/;
  await openMenu(input, { user });
  await type(input, option, { user });

  await select(input, createOptionText, { ...config, user });

  if (waitForElement) {
    await findByText(getReactSelectContainerFromInput(input), option);
  }
}

/**
 * Utility for clearing the first value of a `react-select` dropdown.
 * @param {HTMLElement} input The input field (eg. `getByLabelText('The label')`)
 */
export async function clearFirst(
  input: HTMLElement,
  { user = userEvent }: UserEventOptions = {}
) {
  const container = getReactSelectContainerFromInput(input);
  // The "clear" button is the first svg element that is hidden to screen readers
  const clearButton = container.querySelector('svg[aria-hidden="true"]')!;
  await clear(clearButton, { user });
}

/**
 * Utility for clearing all values in a `react-select` dropdown.
 * @param {HTMLElement} input The input field (eg. `getByLabelText('The label')`)
 */
export async function clearAll(
  input: HTMLElement,
  { user = userEvent }: UserEventOptions = {}
) {
  const container = getReactSelectContainerFromInput(input);
  // The "clear all" button is the penultimate svg element that is hidden to screen readers
  // (the last one is the dropdown arrow)
  const elements = container.querySelectorAll('svg[aria-hidden="true"]');
  const clearAllButton = elements[elements.length - 2];
  await clear(clearAllButton, { user });
}

function setup(user: User): typeof selectEvent {
  return {
    select: (...params: Parameters<typeof select>) =>
      select(params[0], params[1], { user, ...params[2] }),
    create: (...params: Parameters<typeof create>) =>
      create(params[0], params[1], { user, ...params[2] }),
    clearFirst: (...params: Parameters<typeof clearFirst>) =>
      clearFirst(params[0], { user, ...params[1] }),
    clearAll: (...params: Parameters<typeof clearAll>) =>
      clearAll(params[0], { user, ...params[1] }),
    openMenu: (...params: Parameters<typeof openMenu>) =>
      openMenu(params[0], { user, ...params[1] }),
  };
}

const selectEvent = { select, create, clearFirst, clearAll, openMenu };
export default { ...selectEvent, setup };
