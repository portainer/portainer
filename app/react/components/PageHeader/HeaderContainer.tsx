import { PropsWithChildren, createContext, useContext } from 'react';
import clsx from 'clsx';

import styles from './HeaderContainer.module.css';

const Context = createContext<null | boolean>(null);
Context.displayName = 'PageHeaderContext';

export function useHeaderContext() {
  const context = useContext(Context);

  if (context == null) {
    throw new Error('Should be nested inside a HeaderContainer component');
  }
}
interface Props {
  id?: string;
}

export function HeaderContainer({ id, children }: PropsWithChildren<Props>) {
  return (
    <Context.Provider value>
      <div id={id} className={clsx('row', styles.header)}>
        <div id="loadingbar-placeholder" />
        <div className="col-xs-12">
          <div className={styles.meta}>{children}</div>
        </div>
      </div>
    </Context.Provider>
  );
}
