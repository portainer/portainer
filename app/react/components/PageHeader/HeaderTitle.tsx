import { PropsWithChildren } from 'react';

import { useUser } from '@/portainer/hooks/useUser';

import { useHeaderContext } from './HeaderContainer';

interface Props {
  title: string;
}

export function HeaderTitle({ title, children }: PropsWithChildren<Props>) {
  useHeaderContext();
  const { user } = useUser();

  return (
    <div className="page white-space-normal">
      {title}
      <span className="header_title_content">{children}</span>
      {user && !window.ddExtension && (
        <span className="pull-right user-box">
          <i className="fa fa-user-circle" aria-hidden="true" />
          {user.Username}
        </span>
      )}
    </div>
  );
}
