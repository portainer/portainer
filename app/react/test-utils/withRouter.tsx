import { ComponentType } from 'react';
import {
  ReactStateDeclaration,
  UIRouter,
  UIRouterReact,
  UIView,
  hashLocationPlugin,
  servicesPlugin,
} from '@uirouter/react';

/**
 * A helper function to wrap a component with a UIRouter Provider.
 *
 * should only be used in tests
 */
export function withTestRouter<T>(
  WrappedComponent: ComponentType<T>,
  {
    route = '/',
    stateConfig = [],
  }: { route?: string; stateConfig?: Array<ReactStateDeclaration> } = {}
): ComponentType<T> {
  const router = new UIRouterReact();

  // router.trace.enable(Category.TRANSITION);
  router.plugin(servicesPlugin);
  router.plugin(hashLocationPlugin);

  // Set up your custom state configuration
  stateConfig.forEach((state) => router.stateRegistry.register(state));
  router.urlService.rules.initial({ state: route });

  // Try to create a nice displayName for React Dev Tools.
  const displayName =
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

  function WrapperComponent(props: T & JSX.IntrinsicAttributes) {
    return (
      <UIRouter router={router}>
        <UIView />
        <WrappedComponent {...props} />
      </UIRouter>
    );
  }

  WrapperComponent.displayName = `withTestRouter(${displayName})`;

  return WrapperComponent;
}
