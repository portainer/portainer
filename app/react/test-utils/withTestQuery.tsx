import { ComponentType } from 'react';
import { QueryClient } from 'react-query';

import { withReactQuery } from '@/react-tools/withReactQuery';

export function withTestQueryProvider<T>(
  WrappedComponent: ComponentType<T & JSX.IntrinsicAttributes>
) {
  const testQueryClient = new QueryClient({
    defaultOptions: { queries: { retry: false } },
  });

  return withReactQuery(WrappedComponent, testQueryClient);
}
