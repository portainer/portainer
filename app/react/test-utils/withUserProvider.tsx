import { ComponentType, useMemo } from 'react';

import { UserContext } from '@/react/hooks/useUser';
import { User } from '@/portainer/users/types';

const mockUser: User = {
  EndpointAuthorizations: [],
  Id: 1,
  Role: 1,
  Username: 'mock',
  ThemeSettings: {
    color: 'auto',
  },
};

export function withUserProvider<T>(
  WrappedComponent: ComponentType<T>,
  user = mockUser
): ComponentType<T> {
  // Try to create a nice displayName for React Dev Tools.
  const displayName =
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

  function WrapperComponent(props: T & JSX.IntrinsicAttributes) {
    const state = useMemo(() => ({ user }), []);

    return (
      <UserContext.Provider value={state}>
        {/* eslint-disable-next-line react/jsx-props-no-spreading */}
        <WrappedComponent {...props} />
      </UserContext.Provider>
    );
  }

  WrapperComponent.displayName = `withUserProvider(${displayName})`;

  return WrapperComponent;
}
