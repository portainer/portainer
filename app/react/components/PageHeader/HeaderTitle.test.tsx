import { render } from '@testing-library/react';

import { UserViewModel } from '@/portainer/models/user';
import { withUserProvider } from '@/react/test-utils/withUserProvider';
import { withTestRouter } from '@/react/test-utils/withRouter';
import { withTestQueryProvider } from '@/react/test-utils/withTestQuery';

import { HeaderContainer } from './HeaderContainer';
import { HeaderTitle } from './HeaderTitle';

test('should not render without a wrapping HeaderContainer', async () => {
  const consoleErrorFn = vi
    .spyOn(console, 'error')
    .mockImplementation(() => vi.fn());

  const title = 'title';
  function renderComponent() {
    const Wrapped = withTestQueryProvider(HeaderTitle);
    return render(<Wrapped title={title} />);
  }

  expect(renderComponent).toThrowErrorMatchingSnapshot();

  consoleErrorFn.mockRestore();
});

test('should display a HeaderTitle', async () => {
  const username = 'username';
  const user = new UserViewModel({ Username: username });

  const title = 'title';

  const Wrapped = withTestQueryProvider(
    withUserProvider(
      withTestRouter(() => (
        <HeaderContainer>
          <HeaderTitle title={title} />
        </HeaderContainer>
      )),
      user
    )
  );

  const { queryByText } = render(<Wrapped />);

  const heading = queryByText(title);
  expect(heading).toBeVisible();

  expect(queryByText(username)).toBeVisible();
});
