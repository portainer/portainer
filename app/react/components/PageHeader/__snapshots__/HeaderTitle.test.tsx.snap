// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should not render without a wrapping HeaderContainer 1`] = `[Error: Should be nested inside a HeaderContainer component]`;
